### PR TITLE
Support future-proof enums as enums

### PR DESCRIPTION
**Pull request by ertrzyiks**
## Description

Related #6061

## Type of change

Please delete options that are not relevant.

- n/a Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- n/a Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Screenshots/Sandbox (if appropriate/relevant):

Adding links to sandbox or providing screenshots can help us understand more about this PR and take action on it as appropriate

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Environment**:
- OS: 
- `@graphql-codegen/...`: 
- NodeJS: 

## Checklist:

- [ ] I have followed the [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by dotansimha** [comment]
This is great, thanks @ertrzyiks . Feel free to ping me if you need any help!

**Comment by ertrzyiks** [comment]
I will push it forward in the next weeks. I'm going to start vacations so I should have more time for it than now.

**Comment by ertrzyiks** [comment]
@dotansimha it's ready for an early feedback. I would love to play with it a bit more though, but from what I see the canary version is not available to forks (very reasonable though).

To sum up the changes:
 1. keep `futureProofEnums` working for `enumsAsTypes` as it used to, the future value is added to the enum type
 2. make `futureProofEnums` work for other types by changing the definition of enum usages
 3. there already was an option tweaking the usage configured with `allowEnumStringTypes`, so to make it work with both flags I added an extra code branch for that case

File: packages/plugins/typescript/typescript/src/config.ts [hunk 1]
```diff
@@ -84,7 +84,7 @@ export interface TypeScriptPluginConfig extends RawTypesConfig {
    */
   numericEnums?: boolean;
   /**
-   * @description This option controls whether or not a catch-all entry is added to enum type definitions for values that may be added in the future. You also have to set `enumsAsTypes` to true if you wish to use this option.
```
Comment (ertrzyiks) [inline packages/plugins/typescript/typescript/src/config.ts:87]: The removed sentence is no longer true. Should we go into details how this behavior is different for `enumsAsTypes` though?

**Comment by ertrzyiks** [comment]
@dotansimha can I do anything to push this PR forward?

**Review by n1ru4l — APPROVED** [review]

**Comment by n1ru4l** [comment]
@ertrzyiks Thank you 🥳

**Pull request merged**